### PR TITLE
[TTAHUB-2139/2121] Share my groups

### PR DESCRIPTION
**Pull request by GarrettEHill**
## Description of change

In this PR we add both the FE and BE for share my groups. This change includes new sections for groups on the 'Account Management' page. As well as the ability to share groups with co-owners (editors) and individuals on the edit group page. We also update the view group page to show creator, recipeints, coowners, and individuals.

image

image

image

## How to test

- Review the code
- Create a group that is shared with both co-owners and individuals.
- VERIFY: All users in the co-owner and indvidual list have access to (read, write, approve) reports in the exact same regions.
- Make sure you can make updates to that group.
- Impersonate both the co-owner and individuals you should see that group in the appropriate section.
- Make sure as a co-owner you can edit and as an individual you can only view.
- Make sure as both roles you can assign the group to an AR (respectfully).
- Make sure you can still create a private group that is NOT shared with anyone.
- View a group and make sure the readonly view shows the correct values.

**Note:** If you create a group with region 2 grants and your user defaults to creating grants in region 1 you will NOT see the group.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2139

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions
- [x] UI review complete

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [x] Update JIRA ticket status

### CODE REVIEW

File: src/services/groups.ts
Comment (thewatermethod) [inline src/services/groups.ts:726]:
```suggestion
/**
 * Determine the grants, co-owners, and shareWiths to add and remove
 */

// Find grants that are not already existing group grants
const addGrants = grants
  .filter((grantId) => !existingGroupGrants
    .find((egg) => egg.dataValues.grantId === grantId));

// Find existing group grants that are not in the grants list
const removeGrants = existingGroupGrants
  .filter(({
    dataValues: { grantId },
  }: {
    dataValues: { grantId: number },
  }) => !grants.find((grant: number) => grant === grantId))
  .map(({
    dataValues: { grantId },
  }: {
    dataValues: { grantId: number },
  }) => grantId);

// Find co-owners that are not already current group co-owners
const addCoOwners = coOwners
  .filter(({ userId }) => !currentGroupCoOwners
    .find((cgco) => cgco.dataValues.userId === userId)),;

// Find current group co-owners that are not in the co-owners list
const removeCoOwners = currentGroupCoOwners
  .filter(({
    dataValues: { userId },
  }: {
    dataValues: { userId: number },
  }) => !coOwners.find(({ userId: user }) => user === userId))
  .map(({
    dataValues: { userId },
  }: {
    dataValues: { userId: number },
  }) => userId);

// Find shareWiths that are not already current group shareWiths
const addShareWiths = shareWiths
  .filter(({ userId }) => !currentGroupShareWiths
    .find((cgc) => cgc.dataValues.userId === userId));

// Find current group shareWiths that are not in the shareWiths list
const removeShareWiths = currentGroupShareWiths
  .filter(({
    dataValues: { userId },
  }: {
    dataValues: { userId: number },
  }) => !coOwners.find(({ userId: user }) => user === userId))
  .map(({
    dataValues: { userId },
  }: {
    dataValues: { userId: number },
  }) => userId);
  
```

suggestion for readability

File: src/services/groups.ts
Comment (AdamAdHocTeam) [inline src/services/groups.ts:46]: @GarrettEHill I'm starting to update this code. What is the idea here for GroupGrants are we also creating the userId column in that table? What about group types?

Comment (AdamAdHocTeam) [reply src/services/groups.ts:46]: So after looking at this again I guess the idea would be we pull from group collaborators then use the 'groupId' to get any linked grants.

File: frontend/src/pages/AccountManagement/components/MyGroup.scss
Comment (AdamAdHocTeam) [inline frontend/src/pages/AccountManagement/components/MyGroup.scss:2]: Each table width is set to USWDS 'maxw-tablet-lg' so to prevent squishing the action items (view, edit, delete) we set a max-width on the first column. We might ultimately need to set this on the Last Updated column.

Comment (AdamAdHocTeam) [reply frontend/src/pages/AccountManagement/components/MyGroup.scss]: removed for now.

File: frontend/src/components/MultiSelect.js
Comment (thewatermethod) [inline frontend/src/components/MultiSelect.js:252]:
```suggestion
  rules: PropTypes.shape({
    validate: PropTypes.function
  }),
```

??

Comment (AdamAdHocTeam) [reply frontend/src/components/MultiSelect.js:252]: This was just to clarify what needs to happen in order to add a custom validate.

**Comment by thewatermethod** [comment]
Possibly files committed in error:

files in github file view, committed in error

File: frontend/src/pages/AccountManagement/MyGroups.js
Comment (thewatermethod) [inline frontend/src/pages/AccountManagement/MyGroups.js:96]: considering how many fields are being updated, might make sense to use reset instead of setValue (see how TrainingReportForm does it)

https://github.com/HHS/Head-Start-TTADP/blob/33f7d6757ad982d475870a9aa6d0fda5c26069a3/frontend/src/pages/TrainingReportForm/index.js#L41

Comment (thewatermethod) [reply frontend/src/pages/AccountManagement/MyGroups.js:96]: https://www.react-hook-form.com/api/useform/reset/

**Comment by thewatermethod** [comment]
https://github.com/HHS/Head-Start-TTADP/assets/3288586/feca8a88-bd62-4e08-b3af-dada4e0c8280

If a group is public, but shared with specific individuals, I would think that users who the group _isn't_ shared with would not see the group as "shared with me" - they wouldn't see it at all, and they wouldn't have the option to select it in the filter. Am I misunderstanding this feature?

annotations from figma

File: frontend/src/pages/AccountManagement/MyGroups.js
Comment (thewatermethod) [inline frontend/src/pages/AccountManagement/MyGroups.js:394]: I am able to submit the form without selecting anything for this field

Comment (AdamAdHocTeam) [reply frontend/src/pages/AccountManagement/MyGroups.js:388]: If you are talking about the radio buttons you shouldn't have to select anything but it will default to 'everyone' the next time you edit the group.

I also made a recent change that no longer requires co-owners.

**Comment by AdamAdHocTeam** [comment]
> Screen.Recording.2024-02-26.at.4.24.55.PM.mov 
> If a group is public, but shared with specific individuals, I would think that users who the group _isn't_ shared with would not see the group as "shared with me" - they wouldn't see it at all, and they wouldn't have the option to select it in the filter. Am I misunderstanding this feature?
> 
> annotations from figma

This should be fixed.

File: frontend/src/pages/AccountManagement/MyGroups.js
Comment (thewatermethod) [reply frontend/src/pages/AccountManagement/MyGroups.js:388]: If there's a default - 'everyone' - the radio buttons should have that selected as a default value.

Comment (AdamAdHocTeam) [reply frontend/src/pages/AccountManagement/MyGroups.js:388]: Kelly was very adamant about having neither option selected by default.

Comment (AdamAdHocTeam) [reply frontend/src/pages/AccountManagement/MyGroups.js:388]: But I agree with you.

**Comment by thewatermethod** [comment]
Screenshot 2024-02-27 at 10 39 09 AM

Group is not shared with this user (shared with individuals), but user sees group as a filter option

File: frontend/src/pages/AccountManagement/components/MyGroup.scss
Comment (thewatermethod) [inline frontend/src/pages/AccountManagement/components/MyGroup.scss:3]:
```suggestion
 }
```

File: frontend/src/pages/AccountManagement/components/MyGroup.scss
Comment (thewatermethod) [inline frontend/src/pages/AccountManagement/components/MyGroup.scss:7]:
```suggestion
  }
```

File: src/services/groups.ts
Comment (AdamAdHocTeam) [inline src/services/groups.ts:1061]: Remove await.

**Review by thewatermethod — APPROVED** [review]
Looks good. Let's wait until goal nudge is in prod before we merge this into main.

**Pull request merged**
